vraptor-core/src/test/java/br/com/caelum/vraptor/view/DefaultRefererResultTest.java

```diff
@@ -150,4 +151,24 @@ public void whenRefererMatchesAControllerShouldForwardToIt() throws Exception {
 		verify(logic).forwardTo(RefererController.class);
 		verify(controller).index();
 	}
+	
+	@Test
+	public void getRefererShouldReturnCorrectValue() throws Exception {
+		when(request.getHeader("Referer")).thenReturn("http://vraptor.caelum.com.br/test/anything/ok"); //Regular case
+		when(request.getContextPath()).thenReturn("/test");
+		assertEquals("/anything/ok", refererResult.getReferer());
+
+		when(request.getHeader("Referer")).thenReturn("http://vraptor.caelum.com.br/anything/ok/test"); //URL without context path (it appears in the path but not in in the right place)
+		when(request.getContextPath()).thenReturn("/test");
+		assertEquals("/anything/ok/test", refererResult.getReferer());
+		
+		when(request.getHeader("Referer")).thenReturn("http://vraptor.caelum.com.br/vrap/anything/ok"); //URL host starts with ctx path string (/vrap)
+		when(request.getContextPath()).thenReturn("/vrap");
+		assertEquals("/anything/ok", refererResult.getReferer());
+		
+		when(request.getHeader("Referer")).thenReturn("/vrap/anything/ok/vrap"); //relative path in the referer (not common but predicted by http spec)
+		when(request.getContextPath()).thenReturn("/vrap");
+		assertEquals("/anything/ok/vrap", refererResult.getReferer());
```

Comment:
split this test into 4 `@Test`'s, with the appropriate descriptions as the method name instead of the comments.